File: LeetCode-Swift/LeetCode-Swift/堆/_215_数组中的第K个最大元素.swift
//
//  _215_数组中的第K个最大元素.swift
//  LeetCode-Swift
//
//  Created by warden on 2020/2/23.
//  Copyright © 2020 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/kth-largest-element-in-an-array/

import Foundation

class _215_数组中的第K个最大元素 {
    func findKthLargest(_ nums: [Int], _ k: Int) -> Int {
        var numsArr = nums
        var numSize = nums.count
        for i in stride(from: numSize >> 1 - 1, through: 0, by: -1) {
            siftDown(&numsArr, numSize, i)
        }
        for i in 0..<k {
            let tmp = numsArr[numSize - 1]
            numsArr[numSize - 1] = numsArr[0]
            numsArr[0] = tmp
            numSize -= 1;
            if i < k - 1 {
                siftDown(&numsArr, numSize, 0);
            }
        }
        
        return numsArr[nums.count - k]
    }
    
    func siftDown(_ nums: inout [Int], _ size: Int, _ index: Int) {
        let num = nums[index]
        let half = size >> 1
        var idx = index
        while idx < half {
            var childIdx = (idx << 1) + 1
            var child = nums[childIdx]
            let rightIdx = childIdx + 1
            
            if rightIdx < size && nums[rightIdx] > child {
                child = nums[rightIdx]
                childIdx = rightIdx
            }
            if num >= child {
                break
            }
            nums[idx] = child
            idx = childIdx
        }
        nums[idx] = num
    }
}

/**
示例 1:

输入: [3,2,1,5,6,4] 和 k = 2
输出: 5
 
示例 2:

输入: [3,2,3,1,2,4,5,5,6] 和 k = 4
输出: 4
*/
func test_215() {
    let nums = [3,2,1,5,6,4]
    let k = 2
//    let nums = [3,2,3,1,2,4,5,5,6]
//    let k = 4
    let solution = _215_数组中的第K个最大元素()
    print(solution.findKthLargest(nums, k))
}

File: LeetCode-C++/LeetCode-C++/堆/215.数组中的第K个最大元素.cpp
//
//  215.数组中的第K个最大元素.cpp
//  LeetCode-C++
//
//  Created by warden on 2020/2/23.
//  Copyright © 2020 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/kth-largest-element-in-an-array/

#include <stdio.h>
#include <vector>

using namespace std;

class Solution {
private:
    int numSize;
public:
    int findKthLargest(vector<int>& nums, int k) {
        numSize = (int)nums.size();
        heapify(nums);
        for (int i = 0; i < k; i++) {
            swap(nums[numSize-1], nums[0]);
            numSize--;
            siftDown(nums, 0);
        }
        return nums[nums.size() - k];
    }
    
    /**
     建堆
     */
    void heapify(vector<int>& nums) {
        // 自下而上的下溢
        for (int i = (numSize >> 1) - 1; i >= 0; i--) {
            siftDown(nums, i);
        }
    }
    
    void siftDown(vector<int>& nums, int index) {
        int num = nums[index];
        int half = numSize >> 1;
        // 让第一个叶子节点的索引 == 非叶子节点的数量
        while (index < half) {
            int childIndex = (index << 1) + 1;
            int child = nums[childIndex];
            
            int rightIndex = childIndex + 1;
            
            // 选出左右子节点最大的那个
            if (rightIndex < numSize && nums[rightIndex] > child) {
                child = nums[rightIndex];
                childIndex = rightIndex;
            }
            
            if (num >= child) {
                break;
            }
            nums[index] = child;
            index = childIndex;
        }
        nums[index] = num;
    }
};

/**
 示例 1:

 输入: [3,2,1,5,6,4] 和 k = 2
 输出: 5
 示例 2:

 输入: [3,2,3,1,2,4,5,5,6] 和 k = 4
 输出: 4
 */
/**
int main(int argc, const char * argv[]) {
    Solution solution = Solution();
//    vector<int> nums = {3, 2, 1, 5, 6, 4};
//    int result = solution.findKthLargest(nums, 2);
    vector<int> nums = {3, 2, 3, 1, 2, 4, 5, 5, 6};
    int result = solution.findKthLargest(nums, 4);
    printf("%d\n", result);

    return 0;
}
 */
